File: src/quickentry/QuickEntryController.py
import logging
import threading
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from src.quickentry.QuickEntrySubscriber import QuickEntrySubscriber
    from src.quickentry.FocusTracker import FocusTracker
    from src.quickentry.QuickEntryPopup import QuickEntryPopup
    from src.KeyboardSimulator import KeyboardSimulator
    import tkinter as tk


class QuickEntryController:

    _INSERT_DELAY_MS = 50

    def __init__(
        self,
        subscriber: 'QuickEntrySubscriber',
        focus_tracker: 'FocusTracker',
        keyboard_simulator: 'KeyboardSimulator',
        root: 'tk.Tk',
        verbose: bool = False
    ) -> None:
        self._subscriber = subscriber
        self._focus_tracker = focus_tracker
        self._keyboard = keyboard_simulator
        self._root = root
        self._verbose = verbose

        self._popup: Optional['QuickEntryPopup'] = None
        self._lock = threading.Lock()
        self._is_showing = False

    @property
    def is_showing(self) -> bool:
        with self._lock:
            return self._is_showing

    def on_hotkey(self) -> None:
        # Schedule on main thread to avoid tkinter threading issues
        self._root.after(0, self._toggle_popup)

    def _toggle_popup(self) -> None:
        with self._lock:
            if self._is_showing:
                self._hide_and_cancel_locked()
            else:
                self._show_popup_locked()

    def _show_popup(self) -> None:
        with self._lock:
            self._show_popup_locked()

    def _show_popup_locked(self) -> None:
        self._focus_tracker.save_focus()
        self._subscriber.activate()

        if self._popup is None:
            from src.quickentry.QuickEntryPopup import QuickEntryPopup
            self._popup = QuickEntryPopup(
                self._root,
                on_submit=self._on_submit,
                on_cancel=self._on_cancel,
                verbose=self._verbose
            )

        self._popup.set_text("")
        self._popup.show()
        self._is_showing = True

        if self._verbose:
            logging.info("QuickEntryController: popup shown")

    def _hide_and_cancel_locked(self) -> None:
        self._subscriber.deactivate()

        if self._popup:
            self._popup.hide()

        self._is_showing = False
        self._focus_tracker.restore_focus()

        if self._verbose:
            logging.info("QuickEntryController: popup hidden (cancelled)")

    def _on_submit(self) -> None:
        text = self._subscriber.get_accumulated_text()
        self._subscriber.deactivate()

        with self._lock:
            if self._popup:
                self._popup.hide()
            self._is_showing = False

        self._focus_tracker.restore_focus()

        if text.strip():
            self._root.after(self._INSERT_DELAY_MS, lambda: self._keyboard.type_text(text))

        if self._verbose:
            logging.info(f"QuickEntryController: submitted '{text}'")

    def _on_cancel(self) -> None:
        with self._lock:
            self._hide_and_cancel_locked()

    def on_text_change(self, finalized: str, partial: str) -> None:
        if partial:
            display_text = f"{finalized} {partial}".strip()
        else:
            display_text = finalized

        # Schedule update on main thread
        self._root.after(0, lambda: self._update_display(display_text))

    def _update_display(self, text: str) -> None:
        with self._lock:
            if self._popup and self._is_showing:
                self._popup.set_text(text)

File: src/client/winui/SttClient.Core/QuickEntry/QuickEntryController.cs
using Microsoft.Extensions.Logging;
using SttClient.Insertion;

namespace SttClient.QuickEntry;

/// <summary>
/// Orchestrates the QuickEntry popup lifecycle: show on hotkey, update live text,
/// submit (type text + restore focus), and cancel.
///
/// Responsibilities:
/// - Toggles popup visibility on each hotkey press.
/// - Delegates text accumulation to <see cref="QuickEntrySubscriber"/>.
/// - On submit: types accumulated text via <see cref="IKeyboardSimulator"/> and restores focus.
/// - On cancel: hides popup and restores focus without typing.
/// - Registers Enter/Escape as global hotkeys while the popup is visible so the user
///   can submit or cancel without the popup having keyboard focus.
/// - Thread-safe: <see cref="OnHotkey"/> may be called from the hotkey listener thread.
/// </summary>
public sealed class QuickEntryController
{
    private readonly QuickEntrySubscriber _subscriber;
    private readonly IFocusTracker _focusTracker;
    private readonly IKeyboardSimulator _keyboard;
    private readonly IQuickEntryPopup _popup;
    private readonly IPopupHotkeyRegistrar _hotkeyRegistrar;
    private readonly ILogger<QuickEntryController> _logger;
    private readonly object _lock = new();

    private bool _isShowing;

    /// <summary>
    /// Initializes a new <see cref="QuickEntryController"/>.
    /// </summary>
    /// <param name="subscriber">Accumulates live transcription text for the popup.</param>
    /// <param name="focusTracker">Saves and restores the target application's focus.</param>
    /// <param name="keyboard">Types submitted text into the restored window.</param>
    /// <param name="popup">The popup view (real WinUI window or test fake).</param>
    /// <param name="hotkeyRegistrar">Registers Enter/Escape as global hotkeys while popup is visible.</param>
    /// <param name="logger">Logger for diagnostic output.</param>
    public QuickEntryController(
        QuickEntrySubscriber subscriber,
        IFocusTracker focusTracker,
        IKeyboardSimulator keyboard,
        IQuickEntryPopup popup,
        IPopupHotkeyRegistrar hotkeyRegistrar,
        ILogger<QuickEntryController> logger)
    {
        _subscriber = subscriber;
        _focusTracker = focusTracker;
        _keyboard = keyboard;
        _popup = popup;
        _hotkeyRegistrar = hotkeyRegistrar;
        _logger = logger;

        // Wire the subscriber's text-change callback to update the popup display.
        _subscriber.SetTextChangeCallback(UpdatePopupText);
    }

    /// <summary>Gets whether the popup is currently visible.</summary>
    public bool IsShowing
    {
        get { lock (_lock) { return _isShowing; } }
    }

    /// <summary>
    /// Called by <see cref="GlobalHotkeyListener"/> when the hotkey fires.
    /// Toggles the popup open/closed.
    /// </summary>
    public void OnHotkey()
    {
        lock (_lock)
        {
            if (_isShowing)
                HideAndCancel();
            else
                ShowPopup();
        }
    }

    /// <summary>Submits the accumulated text: types it and hides the popup.</summary>
    public void Submit()
    {
        string text;

        lock (_lock)
        {
            text = _subscriber.GetAccumulatedText();
            ClosePopup();
        }

        if (text.Trim().Length > 0)
        {
            _logger.LogInformation("QuickEntryController: submitting '{Text}'", text);
            // Type while our process still owns the foreground — SendInput is blocked by
            // Windows if the calling process is not the foreground process. Restoring focus
            // to the target window first would surrender foreground and cause SendInput to
            // return 0. We type first, then hand foreground back.
            _keyboard.TypeText(text);
        }

        _focusTracker.RestoreFocus();
    }

    /// <summary>Cancels the QuickEntry session: hides the popup without typing.</summary>
    public void Cancel()
    {
        lock (_lock)
        {
            ClosePopup();
        }

        _focusTracker.RestoreFocus();
        _logger.LogInformation("QuickEntryController: popup hidden (cancelled)");
    }

    private void ShowPopup()
    {
        _focusTracker.SaveFocus();
        _subscriber.Activate();
        _popup.Show(onSubmit: Submit, onCancel: Cancel);
        _hotkeyRegistrar.RegisterPopupHotkeys(onSubmit: Submit, onCancel: Cancel);
        _isShowing = true;
        _logger.LogInformation("QuickEntryController: popup shown");
    }

    private void HideAndCancel()
    {
        ClosePopup();
        _focusTracker.RestoreFocus();
        _logger.LogInformation("QuickEntryController: popup hidden (hotkey toggle)");
    }

    private void ClosePopup()
    {
        _subscriber.Deactivate();
        _popup.Hide();
        _popup.SetText(string.Empty);
        _hotkeyRegistrar.UnregisterPopupHotkeys();
        _isShowing = false;
    }

    private void UpdatePopupText(string text)
    {
        lock (_lock)
        {
            if (_isShowing)
                _popup.SetText(text);
        }
    }
}
